core/src/main/java/ai/timefold/solver/core/impl/domain/solution/cloner/FieldAccessingSolutionCloner.java

```diff
@@ -297,87 +341,152 @@ private <C> C cloneCollectionsElementIfNeeded(C original, Map<Object, Object> or
+            this.solutionDescriptor = solutionDescriptor;
             this.declaringClass = declaringClass;
             this.isDeepCloned = DeepCloningUtils.isClassDeepCloned(solutionDescriptor, declaringClass);
         }
 
+        public MethodHandle getConstructor() {
+            if (constructor == null) {
+                synchronized (this) {
+                    if (constructor == null) { // Double-checked locking
+                        try {
+                            var ctor = declaringClass.getDeclaredConstructor();
+                            ctor.setAccessible(true);
+                            constructor = MethodHandles.lookup()
+                                    .unreflectConstructor(ctor);
+                        } catch (ReflectiveOperationException e) {
+                            throw new IllegalStateException(
+                                    "To create a planning clone, the class (%s) must have a no-arg constructor."
+                                            .formatted(declaringClass.getCanonicalName()),
+                                    e);
+                        }
+                    }
+                }
+            }
+            return constructor;
+        }
+
         public ShallowCloningFieldCloner[] getCopiedFieldArray() {
-            if (copiedFieldArray == null) { // Lazy-loaded; some types (such as String) will never get here.
-                copiedFieldArray = Arrays.stream(declaringClass.getDeclaredFields())
-                        .filter(f -> !Modifier.isStatic(f.getModifiers()))
-                        .filter(field -> DeepCloningUtils.isImmutable(field.getType()))
-                        .peek(f -> {
-                            if (DeepCloningUtils.needsDeepClone(solutionDescriptor, f, declaringClass)) {
-                                throw new IllegalStateException("""
-                                        The field (%s) of class (%s) needs to be deep-cloned,
-                                        but its type (%s) is immutable and can not be deep-cloned.
-                                        Maybe remove the @%s annotation from the field?
-                                        Maybe do not reference planning entities inside Java records?
-                                        """
-                                        .strip()
-                                        .formatted(f.getName(), declaringClass.getCanonicalName(),
-                                                f.getType().getCanonicalName(), DeepPlanningClone.class.getSimpleName()));
-                            } else {
-                                f.setAccessible(true);
-                            }
-                        })
-                        .map(ShallowCloningFieldCloner::of)
-                        .toArray(ShallowCloningFieldCloner[]::new);
+            if (copiedFieldArray == null) {
+                synchronized (this) {
```

Comment:
Do move threads/solver manager share the same solution cloner?

Reply:
Anything that shares the solution descriptor will share the cloner. (There is a 1:1 relationship.) This class was specifically marked as thread-safe before this PR, so I made sure that continues being true.